snuba/subscriptions/executor_consumer.py

```diff
@@ -90,3 +102,127 @@ def create(
         self, commit: Callable[[Mapping[Partition, Position]], None]
     ) -> ProcessingStrategy[KafkaPayload]:
         return Noop(commit)
+
+
+class SubscriptionResultFuture(NamedTuple):
+    message: Message[KafkaPayload]
+    future: Future[Tuple[Request, Result]]
+
+
+class ExecuteQuery(ProcessingStrategy[KafkaPayload]):
+    """
+    Decodes a scheduled subscription task from the Kafka payload, builds
+    the request and executes the ClickHouse query.
+    """
+
+    def __init__(
+        self,
+        dataset: Dataset,
+        max_concurrent_queries: int,
+        metrics: MetricsBackend,
+        next_step: ProcessingStrategy[KafkaPayload],
+    ) -> None:
+        self.__dataset = dataset
+        self.__max_concurrent_queries = max_concurrent_queries
+        self.__metrics = metrics
+        self.__next_step = next_step
+
+        self.__encoder = SubscriptionScheduledTaskEncoder()
+
+        self.__executor = ThreadPoolExecutor()
+        self.__queue: Deque[SubscriptionResultFuture] = deque()
+
+        self.__closed = False
+
+        self.__concurrent_gauge: Gauge = ThreadSafeGauge(
+            self.__metrics, "executor.concurrent"
+        )
+
+        self.__concurrent_clickhouse_gauge: Gauge = ThreadSafeGauge(
+            self.__metrics, "executor.concurrent.clickhouse"
+        )
+
+    def __execute_query(
+        self, task: ScheduledSubscriptionTask, tick: Tick
+    ) -> Tuple[Request, Result]:
+        timer = Timer("query")
+
+        with self.__concurrent_gauge:
+            request = task.task.subscription.data.build_request(
```

Comment:
We will need to track the subscription queries executed by a subscriptions consumer separately from these ones. This way they would have the same referrer and be indistinguishable.
I think we should either ensure the metrics we use to track query timings have a dedicated tag or that we use a different referrer.

Reply:
All the executor metrics are namespaced under `subscriptions.executor` so it's actually a different metric that just follows a similar naming convention to the existing one. (well it should be after this bugfix at least https://github.com/getsentry/snuba/pull/2265)

Reply:
Everything recorded via the Timer will still go into `api` metrics without a distinction between the existing subscriptions and these ones. 
This is going to be a bit tricky, though we should definitively separate the metrics. Either by changing the referrer or by finding a way to add tags to the Timer (maybe changing the referrer is appropriate)

Reply:
Ah I see, thought you were talking about the concurrent gauge

Reply:
Changed the referrer to `subscriptions_executor`. Regular worker is still `subscriptions`.